lib/segment/src/index/hnsw_index/hnsw.rs

```diff
@@ -219,75 +219,83 @@ impl<TGraphLinks: GraphLinks> HNSWIndex<TGraphLinks> {
+        let quantized_storage = vector_storage.quantized_storage();
+
+        // Check that:
+        // - `params` is `Some`
+        // - `params.quantization` is `Some`
+        // - and `params.quantization.ignore` is `false`
+        let quantization_params =
+            params
+                .and_then(|p| p.quantization)
+                .unwrap_or_else(|| QuantizationSearchParams {
+                    ignore: default_quantization_ignore_value(),
+                    rescore: default_quantization_rescore_value(),
+                    oversampling: default_quantization_oversampling_value(),
+                });
+
+        let (raw_scorer, quantized) = match quantized_storage {
+            // If `quantization_params` is `Some`, then quantization is *not* ignored
+            Some(quantized_storage) if !quantization_params.ignore => {
+                let scorer = quantized_storage.raw_scorer(
                     vector,
                     id_tracker.deleted_point_bitslice(),
                     vector_storage.deleted_vector_bitslice(),
-                ),
-                true,
-            )
-        } else {
-            (
-                new_raw_scorer(
+                );
+
+                (scorer, true)
+            }
+
+            _ => {
+                let scorer = new_raw_scorer(
                     vector.to_owned(),
                     &vector_storage,
                     id_tracker.deleted_point_bitslice(),
-                ),
-                false,
-            )
+                );
+
+                (scorer, false)
+            }
         };
-        let payload_index = self.payload_index.borrow();
 
+        let payload_index = self.payload_index.borrow();
         let filter_context = filter.map(|f| payload_index.filter_context(f));
-
         let points_scorer = FilteredScorer::new(raw_scorer.as_ref(), filter_context.as_deref());
 
-        if let Some(graph) = &self.graph {
-            let search_result = graph.search(top, ef, points_scorer);
-            let if_rescore = params
-                .and_then(|p| p.quantization)
-                .map(|q| q.rescore)
-                .unwrap_or(default_quantization_rescore_value());
-            if quantized && if_rescore {
-                let raw_scorer = new_raw_scorer(
-                    vector.to_owned(),
-                    &vector_storage,
-                    id_tracker.deleted_point_bitslice(),
-                );
-                let mut ids_iterator = search_result.iter().map(|x| x.idx);
-                let mut re_scored = raw_scorer.score_points_unfiltered(&mut ids_iterator);
-                re_scored.sort_unstable();
-                re_scored
+        let Some(graph) = &self.graph else {
+            return Vec::new();
+        };
+
+        if quantized && quantization_params.rescore {
+            let oversampling = quantization_params.oversampling.unwrap_or(1.0);
+
+            let oversampled_top = if oversampling > 1.0 {
```

Comment:
maybe oversampling < 1.0 should be a client side error

Reply:
Ideally, it should a validation error. But we don't have validation in search yet

Reply:
Implemented in <https://github.com/qdrant/qdrant/pull/2100>